src/Microsoft.ML.Featurizers/DateTimeTransformer.cs

```diff
@@ -0,0 +1,782 @@
+            }
+        }
+
+        #endregion
+    }
+
+    /// <summary>
+    /// The DateTimeTransformerEstimator splits up a date into all of its sub parts as individual columns. It generates these fields with a user specified prefix:
+    /// int Year, byte Month, byte Day, byte Hour, byte Minute, byte Second, byte AmPm, byte Hour12, byte DayOfWeek, byte DayOfQuarter,
+    /// ushort DayOfYear, ushort WeekOfMonth, byte QuarterOfYear, byte HalfOfYear, byte WeekIso, int YearIso, string MonthLabel, string AmPmLabel,
+    /// string DayOfWeekLabel, string HolidayName, byte IsPaidTimeOff
+    ///
+    /// You can optionally specify a country and it will pull holiday information about the country as well
+    /// </summary>
+    /// <remarks>
+    /// <format type="text/markdown"><![CDATA[
+    ///
+    /// ###  Estimator Characteristics
+    /// |  |  |
+    /// | -- | -- |
+    /// | Does this estimator need to look at the data to train its parameters? | No |
+    /// | Input column data type | Int64 |
+    /// | Output column data type | Columns and types listed in the summary |
+    ///
+    /// The <xref:Microsoft.ML.Transforms.CategoryImputerEstimator> is a trivial estimator and does not need training.
+    ///
+    ///
+    /// ]]>
+    /// </format>
+    /// </remarks>
+    /// <seealso cref="DateTimeTransformerExtensionClass.DateTimeTransformer(TransformsCatalog, string, string, DateTimeTransformerEstimator.ColumnsProduced[])"/>
+    /// <seealso cref="DateTimeTransformerExtensionClass.DateTimeTransformer(TransformsCatalog, string, string, DateTimeTransformerEstimator.ColumnsProduced[], DateTimeTransformerEstimator.Countries)"/>
+    public sealed class DateTimeTransformerEstimator : IEstimator<DateTimeTransformer>
+    {
+        private readonly Options _options;
+
+        private readonly IHost _host;
+
+        #region Options
+        internal sealed class Options : TransformInputBase
+        {
+            [Argument(ArgumentType.Required, HelpText = "Input column", Name = "Source", ShortName = "src", SortOrder = 1)]
+            public string Source;
+
+            // This transformer adds columns
+            [Argument(ArgumentType.Required, HelpText = "Output column prefix", Name = "Prefix", ShortName = "pre", SortOrder = 2)]
+            public string Prefix;
+
+            [Argument(ArgumentType.MultipleUnique, HelpText = "Columns to drop after the DateTime Expansion", Name = "ColumnsToDrop", ShortName = "drop", SortOrder = 3)]
+            public ColumnsProduced[] ColumnsToDrop;
+
+            [Argument(ArgumentType.AtMostOnce, HelpText = "Country to get holidays for. Defaults to none if not passed", Name = "Country", ShortName = "ctry", SortOrder = 4)]
+            public Countries Country = Countries.None;
+        }
+
+        #endregion
+
+        public DateTimeTransformerEstimator(IHostEnvironment env, string inputColumnName, string columnPrefix, ColumnsProduced[] columnsToDrop, Countries country = Countries.None)
```

Comment:
The constructors should all be `internal`. The way these objects are instantiated is through the Catalog methods.

Reply:
Changed to internal.

---
In reply to: [353904541](https://github.com/dotnet/machinelearning/pull/4521#discussion_r353904541) [](ancestors = 353904541)